infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/multiuser/oauth/IdentityProviderConfigBuilder.java

```diff
@@ -0,0 +1,171 @@
+    }
+  }
+
+  /**
+   * Builds the Openshift {@link Config} object based on a default {@link Config} object and an
+   * optional workspace Id.
+   */
+  protected Config buildConfig(Config defaultConfig, @Nullable String workspaceId)
+      throws InfrastructureException {
+    Subject subject = EnvironmentContext.getCurrent().getSubject();
+
+    String workspaceOwnerId = null;
+    if (workspaceId != null) {
+      @SuppressWarnings("rawtypes")
+      Optional<RuntimeContext> context =
+          workspaceRuntimeProvider.get().getRuntimeContext(workspaceId);
+      workspaceOwnerId = context.map(c -> c.getIdentity().getOwnerId()).orElse(null);
+    }
+
+    if (oauthIdentityProvider != null
+        && subject != Subject.ANONYMOUS
+        && (workspaceOwnerId == null || subject.getUserId().equals(workspaceOwnerId))) {
+      try {
+        KeycloakTokenResponse keycloakTokenInfos =
+            keycloakServiceClient.getIdentityProviderToken(oauthIdentityProvider);
+        if ("user:full".equals(keycloakTokenInfos.getScope())) {
+          return new OpenShiftConfigBuilder(OpenShiftConfig.wrap(defaultConfig))
+              .withOauthToken(keycloakTokenInfos.getAccessToken())
+              .build();
+        } else {
+          throw new InfrastructureException(
+              "Cannot retrieve user Openshift token: '"
+                  + oauthIdentityProvider
+                  + "' identity provider is not granted full rights: "
+                  + oauthIdentityProvider);
+        }
+      } catch (UnauthorizedException e) {
+        LOG.error("cannot retrieve User Openshift token from the identity provider", e);
+
+        throw new InfrastructureException(getMessageToLinkAccount());
+      } catch (BadRequestException e) {
+        LOG.error(
+            "cannot retrieve User Openshift token from the '"
+                + oauthIdentityProvider
+                + "' identity provider",
+            e);
+        if (e.getMessage().endsWith("Invalid token.")) {
+          throw new InfrastructureException(
+              "Your session has expired. \nPlease "
+                  + "<a href='javascript:location.reload();' target='_top'>"
+                  + "login"
+                  + "</a> to Che again to get access to your Openshift account");
+        }
+        throw new InfrastructureException(e.getMessage(), e);
+      } catch (Exception e) {
+        LOG.error(
+            "cannot retrieve User Openshift token from the  '"
+                + oauthIdentityProvider
+                + "' identity provider",
+            e);
+        throw new InfrastructureException(e.getMessage(), e);
+      }
+    }
+    return defaultConfig;
+  }
+
+  private String getMessageToLinkAccount() {
+    String referrer_uri =
+        rootUrl.replace("http://", "http%3A%2F%2F").replace("https://", "https%3A%2F%2F")
+            + "%2Fdashboard%2F?redirect_fragment%3D%2Fworkspaces";
+
+    return "You should link your account with the <strong>"
```

Comment:
This is a string that can be generated once in the constructor instead of doing all these string manipulations on each call of the method. Can you move evaluation to the constructor?